sourmash/cli/__init__.py

```diff
@@ -0,0 +1,125 @@
+"""Define the top-level command line interface for sourmash
+
+This module handles user input when sourmash is invoked from the command line.
+A top-level parser is defined for the `sourmash` command, and subparsers are
+defined for each subcommand. Some sourmash operations are grouped together
+using the `sourmash <subcmd> <subsubcmd>` pattern, and these are organized in
+their own CLI submodules, each with a dedicated directory.
+"""
+
+from argparse import ArgumentParser, RawDescriptionHelpFormatter, SUPPRESS
+import os
+import sys
+
+import sourmash
+
+from . import utils
+
+# Commands
+from . import categorize
+from . import compare
+from . import compute
+from . import dump
+from . import gather
+from . import import_csv
+from . import info
+from . import index
+from . import migrate
+from . import multigather
+from . import plot
+from . import sbt_combine
+from . import search
+from . import watch
+
+# Subcommand groups
+from . import lca
+from . import sig
+from . import storage
+
+
+class SourmashParser(ArgumentParser):
+    _citation_printed = False
+
+    def __init__(self, citation=True, **kwargs):
+        super(SourmashParser, self).__init__(**kwargs)
+        self.citation = citation
+
+    @classmethod
+    def print_citation(cls):
```

Comment:
I made `print_citation` a class method, and storing the `_citation_printed` at the class level let's us assert that it is only printed once along all commands (because all commands are subclasses of `SourmashParser`).

(the `classmethod` trick is straight out of [snakemake], btw)

[snakemake]: https://github.com/snakemake/snakemake/blob/e4b1e9a41ddf0ed260e3048586ca488b7fa4a148/snakemake/shell.py